net: if: Improve interface status reporting

rlubos opened the pull request:
Extend interface state management as described in https://www.kernel.org/doc/html/latest/networking/operstates.html and https://www.rfc-editor.org/rfc/rfc2863.

With those changes, `net_if_up()`/`net_if_down()` are only supposed to be called by the application layer, not drivers/L2s. The existing codebase was refactored to reflect that.

Drivers/L2s have not two new flags to reflect their status - `NET_IF_LOWER_UP` and `NET_IF_DORMANT`, along with the helper functions. The combination of those two, along with the administrative state of the interface is translated into the operational state, which indicates the interface's readiness to accept application data.

Resolves #28998

rlubos committed:
net: if: Introduce carrier and dormant management on a network iface

Introduce a new interface state management scheme, according to
RFC 2863. This includes the following changes:

* Introduce a new interface flag: NET_IF_LOWER_UP, along with
  corresponding helper functions. The flag should be set/cleared on an
  interface by a network driver/L2 to signalize physical readiness of an
  interface to transmit data (for example cable plugged in).

* Introduce a new interface flag: NET_IF_DORMANT, along with
  corresponding helper functions. The flag should be set on an
  iterface when the interface is not ready to transmit application data,
  for example still not joined a Wi-Fi network.

* Introduce a new interface flag: NET_IF_RUNNING, indicating that
  interface is ready to transmit application data.

* Update the meaning of the NET_IF_UP flag - it now singnalizes whether
  an interface has been brought up/down by the application (admin
  up/down).

* Introduce operational state of an interface, derived from above. It
  reflects the internal interface state.

The meaning of net_if_is_up() function and NET_EVENT_IF_UP/DOWN events
remains unchanged to retain backward compability - they reflect the
interface readiness to transmit application data.

To verify the administrative up/down state, a new function
net_if_is_admin_up() has been introduced, along with
NET_EVENT_IF_ADMIN_UP/DOWN events.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: if: Fix if down handling for all offloaded interfaces

Socket offloading was not taken into consideration when network
interface was put down. Fix this by introducing a helper function, to
verify whether an interface is offloaded or not.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ethernet: Align Ethernet drivers/L2 with iface state upgrade

Align Ethernet/Wi-Fi drivers/L2 with interface state handling update.

For drivers, that did not support carrier detection, no changes are
needed.

Driver that did support carrier detection, are updated to set the
carrier state to OFF by default, instead of setting the
NET_IF_NO_AUTO_START flag. This allows to postopne the actual
NET_EVENT_IF_UP notification until driver detects that carrier is ready.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: bluetooth: Align Bluetooth L2 with iface state upgrade

Align Bluetooth L2 with interface state handling update. Use the dormant
flag to indicate whether interface has a Bluetooth connection or not.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: openthread: Align OpenThread L2 with iface state upgrade

Align OpenThread L2 with interface state handling update. Use the
dormant flag to indicate whether an interface joined a Thread network.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: netusb: Align netusb driver with iface state upgrade

Align netusb driver with interface state handling update. Instead of
bringing the interface up/down from the driver level, use carrier on/off
notification to update the inteface state.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ppp: Align PPP driver/L2 with iface state upgrade

Align PPP drivers/L2 with interface state handling update. Use the
carrier on/off notification instead of bringing the interface up/down to
update the interface state.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
drivers: modem: hl7800: Align driver with iface status upgrade

Align the offloaded hl7800 driver with interface state handling update.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
drivers: modem: wncm14a2a: Align driver with iface status upgrade

Align the offloaded wncm14a2a driver with interface state handling
update.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
drivers: wifi: esp_at: Align driver with iface status upgrade

Align the offloaded esp_at driver with interface state handling update.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: if: Verify actual interface state instead of admin state on TX

When packet is passed to transmit at the net_if level, verify the
carrier state (NET_IF_LOWER_UP) instead of administrative state
(NET_IF_UP).

The administrative state is checked anyway at higher level (net_context)
so no need to verify it again. This will allow to still transmit control
packets for example when interface is in a dormant state.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: virtual: Align virtual  L2 with iface status upgrade

Align virtual L2 with interface state handling update. Introduce
net_virtual_enable() function, which gets called whenever a network
interface is brought up (operational). This, combined with already
existing net_virtual_disable() function, can be used to update the
carrier state on the virtual interface, based on the underlying
interface status.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: virtual: Add test to verify behavior when real iface goes down

Add test that verifies if a virtual interface state is updated properly
when an underlying real interface is brought down/up.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: if: Remove net_if_carrier_down function

The function is no longer used, as its functionality has been replaced
with net_if_carrier_off().

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
doc: net: Add documentation on network interface admin/oper states

Extend the net_if documentation with a section describing the
administrative and operational states on an interface, and how do they
interact with each other.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
From all of the reworked drivers/L2s, I'm not 100% confident about the PPP - I think the changes I've made make sense when it comes to carrier state management, but I have an impression the PPP interface should also change the dormant state as well. I just didn't feel confident enough in the area to implement this - maybe someone could help/assist with that?

rlubos commented:
CC @krish2718

jukkar commented on include/zephyr/net/net_if.h:
```diff
@@ -193,12 +194,32 @@ enum net_if_flag {
 };
 
+/** Network interface operational status (RFC 2863). */
+enum net_if_oper_state {
+	NET_IF_OPER_UNKNOWN,
+	NET_IF_OPER_NOTPRESENT,
+	NET_IF_OPER_DOWN,
+	NET_IF_OPER_LOWERLAYERDOWN,
+	NET_IF_OPER_TESTING,
+	NET_IF_OPER_DORMANT,
+	NET_IF_OPER_UP,
+};
```
If we set this to
```
enum net_if_oper_state {
....
} __packed;
```
then the `oper_state` variable below can be set to `enum net_if_oper_state oper_state` and it would consume only one byte.

jukkar commented on include/zephyr/net/net_if.h:
```diff
@@ -576,6 +600,34 @@ static inline bool net_if_flag_is_set(struct net_if *iface,
+ * @param iface Pointer to network interface
+ * @param oper_state Operational state to set
+ */
+static inline void net_if_oper_state_set(struct net_if *iface,
+					 enum net_if_oper_state oper_state)
+{
+	NET_ASSERT(iface);
+
+	iface->if_dev->oper_state = oper_state;
+}
```
Minor thing, but I would suggest that we check the value so that it is within allowed range.

jukkar replied:
We could also return the new value so that we could catch any errors. So if the returned value != oper_state, then the value was invalid. Perhaps all this checking is far fetched but it would be nice to catch all these errors.

rlubos replied:
Ok, I've updated `net_if_oper_state_set()` and added a simple error checking.

jukkar commented on include/zephyr/net/net_if.h:
```diff
@@ -2113,6 +2166,90 @@ static inline bool net_if_is_up(struct net_if *iface)
 int net_if_down(struct net_if *iface);
 
+/**
+ * @brief Check if interface was brought up by the application or system.
```
Perhaps better to say
`Check if interface was brought up by the administrator.`
I think we should not say "system" here.

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -4023,6 +4026,104 @@ static inline bool is_iface_offloaded(struct net_if *iface)
+{
+	net_if_flag_set(iface, NET_IF_RUNNING);
+	net_mgmt_event_notify(NET_EVENT_IF_UP, iface);
+
```
Please remove extra empty line

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -4023,6 +4026,104 @@ static inline bool is_iface_offloaded(struct net_if *iface)
+
+static inline const char *net_if_oper_state2str(enum net_if_oper_state state)
+{
+	switch (state) {
+	case NET_IF_OPER_UNKNOWN:
+		return "UNKNOWN";
+	case NET_IF_OPER_NOTPRESENT:
+		return "NOTPRESENT";
+	case NET_IF_OPER_DOWN:
+		return "DOWN";
+	case NET_IF_OPER_LOWERLAYERDOWN:
+		return "LOWERLAYERDOWN";
+	case NET_IF_OPER_TESTING:
+		return "TESTING";
+	case NET_IF_OPER_DORMANT:
+		return "DORMANT";
+	case NET_IF_OPER_UP:
+		return "UP";
+	default:
+		break;
+	}
+
+	return "<invalid>";
```
Perhaps return string values only if debugging is enabled?
If not then, we could just return empty value ""

jukkar reviewed (commented):
Looks good, minor comments added.
The virtual tests were added which is good, but I think we need more tests for this.

rlubos commented:
Thank you @jukkar for the review, I've addressed all of your comments.

> The virtual tests were added which is good, but I think we need more tests for this.

I was thinking on adding some tests to verify operational state transitions on some fake interface, do you have any other specific tests in mind?

jukkar commented:
> I was thinking on adding some tests to verify operational state transitions on some fake interface, do you have any other specific tests in mind?

That sounds good, we can add more later if needed.

jukkar commented on drivers/net/ppp.c:
```diff
@@ -954,14 +954,14 @@ static void ppp_iface_init(struct net_if *iface)
-	 * after the modem is ready or when manually started.
+	/* If the interface autostart is disabled from Kconfig, then do not
+	 * start the interface automatically but only after manually started.
 	 */
-	if (IS_ENABLED(CONFIG_MODEM_GSM_PPP) ||
-	    IS_ENABLED(CONFIG_PPP_NET_IF_NO_AUTO_START)) {
+	if (IS_ENABLED(CONFIG_PPP_NET_IF_NO_AUTO_START)) {
 		net_if_flag_set(iface, NET_IF_NO_AUTO_START);
 	}
+
+	net_if_carrier_off(iface);
```
This might need some more testing, I think there was some reason for extra check but do not recall now what it was. Unfortunately I do not have any GSM modem that I could try atm, if any other reviewer has one, please test this with GSM modem.

rlubos replied:
Unfortunately, I don't have a modem either to verify. But the situation here is similar to what Ethernet drivers did - they disabled interface autostart only to bring it up with `net_if_up()` once the carrier was detected. With those changes, it's no longer needed, as drivers can signalize the carrier state with `net_if_carrier_*()` API. The interface can be brought admin up during system boot, it's only needed that the carrier state is set to off during init, so that `NET_EVENT_IF_UP` is not signified prematurely.

rlubos committed:
tests: net: iface: Add test to verify operational state transitions

Add test which verifies that interface operational state changes
correctly when carrier and dormant status is updated.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
> > I was thinking on adding some tests to verify operational state transitions on some fake interface, do you have any other specific tests in mind?
> 
> That sounds good, we can add more later if needed.

I have added an additional test case to cover that.

jukkar commented on tests/net/iface/src/main.c:
```diff
@@ -435,6 +435,80 @@ static void iface_teardown(void *dummy)
+	test_iface_init(iface1, true, false);
+	zassert_true(net_if_is_up(iface1), "Interface should be up");
+}
+
```
Perhaps I missed it but I think we need a test case where the admin sets the interface down, but then operational state causes the interface to be "ready". The interface should stay down in this case.

rlubos replied:
Yes, such a scenario was not present, I've added it.

jukkar reviewed (approved):
LGTM

cfriedt commented:
@rerickson1 - can you do some modem testing with this?

rerickson1 commented:
> @rerickson1 - can you do some modem testing with this?

Yeah, this looks good. Not sure when I'll get to trying this.
